packages/react-dom-bindings/src/client/ReactDOMFloatClient.js

```diff
@@ -865,135 +906,141 @@ export function hydrateHoistable(
   } else {
     childString = '';
   }
-  nodeLoop: for (let i = 0; i < nodes.length; i++) {
-    const node = nodes[i];
+
+  const attributes: Array<string> = [];
+
+  for (const propName in props) {
+    const propValue = props[propName];
+    if (!props.hasOwnProperty(propName)) {
+      continue;
+    }
     if (
-      isMarkedResource(node) ||
-      node.namespaceURI === SVG_NAMESPACE ||
-      node.textContent !== childString
+      // shouldIgnoreAttribute
+      // We have already filtered out null/undefined and reserved words.
+      propName.length > 2 &&
+      (propName[0] === 'o' || propName[0] === 'O') &&
+      (propName[1] === 'n' || propName[1] === 'N')
     ) {
       continue;
     }
-    let checkedAttributes = 0;
-    for (const propName in props) {
-      const propValue = props[propName];
-      if (!props.hasOwnProperty(propName)) {
+    switch (propName) {
+      // Reserved props will never have an attribute partner
+      case 'children':
+      case 'defaultValue':
+      case 'dangerouslySetInnerHTML':
+      case 'defaultChecked':
+      case 'innerHTML':
+      case 'suppressContentEditableWarning':
+      case 'suppressHydrationWarning':
+      case 'style':
+        // we advance to the next prop
         continue;
+
+      // Name remapped props used by hoistable tag types
+      case 'className': {
+        if (__DEV__) {
+          checkAttributeStringCoercion(propValue, propName);
+        }
+        attributes.push('class', '' + propValue);
+        break;
       }
-      switch (propName) {
-        // Reserved props will never have an attribute partner
-        case 'children':
-        case 'defaultValue':
-        case 'dangerouslySetInnerHTML':
-        case 'defaultChecked':
-        case 'innerHTML':
-        case 'suppressContentEditableWarning':
-        case 'suppressHydrationWarning':
-        case 'style':
-          // we advance to the next prop
-          continue;
-
-        // Name remapped props used by hoistable tag types
-        case 'className': {
-          if (__DEV__) {
-            checkAttributeStringCoercion(propValue, propName);
-          }
-          if (node.getAttribute('class') !== '' + propValue) continue nodeLoop;
-          break;
+      case 'httpEquiv': {
+        if (__DEV__) {
+          checkAttributeStringCoercion(propValue, propName);
         }
-        case 'httpEquiv': {
-          if (__DEV__) {
-            checkAttributeStringCoercion(propValue, propName);
-          }
-          if (node.getAttribute('http-equiv') !== '' + propValue)
-            continue nodeLoop;
-          break;
+        attributes.push('http-equiv', '' + propValue);
+        break;
+      }
+
+      // Boolean props used by hoistable tag types
+      case 'async':
+      case 'defer':
+      case 'disabled':
+      case 'hidden':
+      case 'noModule':
+      case 'scoped':
+      case 'itemScope':
+        if (propValue === true) {
+          attributes.push(propName, '');
```

Comment:
I believe we always enforce that boolean attributes serialize as empty string. If this changes then I may need to switch to hasAttribute but I believe this works for now